File: Data-Structures-Algorithms-In-TS/AlgorithmsChapters/BT/Permutations.ts

function swap(arr: number[], i: number, j: number): void {
    const temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

function permutation(arr: number[], i: number, length: number): void {
    if (length === i) {
        console.log(arr);
        return;
    }

    for (let j = i; j < length; j++) {
        swap(arr, i, j);
        permutation(arr, i + 1, length);
        swap(arr, i, j);
    }
}

function isValid(arr: number[], n: number): boolean {
    for (let j = 1; j < n; j++) {
        if (Math.abs(arr[j] - arr[j - 1]) < 2) {
            return false;
        }
    }
    return true;
}

function permutation2(arr: number[], i: number, length: number): void {
    if (length === i) {
        if (isValid(arr, length)) {
            console.log(arr);
        }
        return;
    }

    for (let j = i; j < length; j++) {
        swap(arr, i, j);
        permutation2(arr, i + 1, length);
        swap(arr, i, j);
    }
}

function isValid2(arr: number[], i: number): boolean {
    if (i < 1 || Math.abs(arr[i] - arr[i - 1]) >= 2) {
        return true;
    }
    return false;
}

function permutation3(arr: number[], i: number, length: number): void {
    if (length === i) {
        console.log(arr);
        return;
    }

    for (let j = i; j < length; j++) {
        swap(arr, i, j);
        if (isValid2(arr, i)) {
            permutation3(arr, i + 1, length);
        }
        swap(arr, i, j);
    }
}

/* Testing code */
const arr: number[] = [1, 2, 3, 4];
permutation(arr, 0, 4);
console.log();
permutation2(arr, 0, 4);
console.log();
permutation3(arr, 0, 4);


/*
[ 1, 2, 3, 4 ]
[ 1, 2, 4, 3 ]
...
[ 4, 1, 3, 2 ]
[ 4, 1, 2, 3 ]
*/
/*
[ 2, 4, 1, 3 ]
[ 3, 1, 4, 2 ]

[ 2, 4, 1, 3 ]
[ 3, 1, 4, 2 ]

*/

File: Data-Structures-and-Algorithms-using-Swift/AlgorithmsChapters/BT/Permutations.swift
import Foundation

func permutation(_ arr : inout [Int], _ i : Int, _ length : Int) {
	if (length == i) {
		print(arr)
		return
	}
	var j : Int = i
	while (j < length) {
		arr.swapAt(i,j)
		permutation( &arr,i + 1,length)
		arr.swapAt(i,j)
		j += 1
	}
	return
}
// 	1 2 3 4 
// 	1 2 4 3 
// 	.....
// 	4 1 3 2 
// 	4 1 2 3

func isValid(_ arr : inout [Int], _ n : Int) -> Bool {
	var j : Int = 1
	while (j < n) {
		if (abs(arr[j] - arr[j - 1]) < 2) {
			return false
		}
		j += 1
	}
	return true
}

func permutation2(_ arr : inout [Int], _ i : Int, _ length : Int) {
	if (length == i) {
		if (isValid( &arr,length)) {
			print(arr)
		}
		return
	}

	var j : Int = i
	while (j < length) {
		arr.swapAt(i,j)
		permutation2( &arr,i + 1,length)
		arr.swapAt(i,j)
		j += 1
	}
	return
}

func isValid2(_ arr : inout [Int], _ i : Int) -> Bool {
	if (i < 1 || abs(arr[i] - arr[i - 1]) >= 2) {
		return true
	}
	return false
}

func permutation3(_ arr : inout [Int], _ i : Int, _ length : Int) {
	if (length == i) {
		print(arr)
		return
	}
	var j : Int = i
	while (j < length) {
		arr.swapAt(i,j)
		if (isValid2( &arr,i)) {permutation3( &arr,i + 1,length)
		}
		arr.swapAt(i,j)
		j += 1
	}
	return
}
	
// Testing code
var arr : [Int] = [1, 2, 3, 4]
permutation( &arr,0,4)
print()
permutation2( &arr,0,4)
print()
permutation3( &arr,0,4)

/* 
[1, 2, 3, 4]
[1, 2, 4, 3]
...
[4, 1, 3, 2]
[4, 1, 2, 3]

[2, 4, 1, 3]
[3, 1, 4, 2]

[2, 4, 1, 3]
[3, 1, 4, 2]
 */